pkg/agent/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,859 @@
+			// use rate limiter to reduce the times we need to update status.
+			if reachTarget || captureState.updateRateLimiter.Allow() {
+				pc, err := c.packetCaptureLister.Get(captureState.name)
+				if err != nil {
+					return fmt.Errorf("get PacketCapture failed: %w", err)
+				}
+				// if reach the target. flush the file and upload it.
+				if reachTarget {
+					path := os.Getenv("POD_NAME") + ":" + nameToPath(pc.Name)
+					if err = captureState.pcapngWriter.Flush(); err != nil {
+						return err
+					}
+					if pc.Spec.FileServer != nil {
+						err = c.uploadPackets(pc, captureState.pcapngFile)
+						klog.V(4).InfoS("Upload captured packets", "name", pc.Name, "path", path)
+
+					}
+					// update capture result.
+					if updateErr := c.updatePacketCaptureStatus(pc.Name, captureState.capturedPacketsNum, path, err); updateErr != nil {
+						klog.ErrorS(err, "Failed to update PacketCapture status")
+					}
+					if err != nil {
+						return err
+					}
+					if err := captureState.pcapngFile.Close(); err != nil {
+						klog.ErrorS(err, "Close pcapng file error", "name", pc.Name, "path", path)
+					}
+				}
+
+				err = c.updatePacketCaptureStatus(pc.Name, captureState.capturedPacketsNum, "", nil)
+				if err != nil {
+					klog.ErrorS(err, "Failed to update PacketCapture status")
+				} else {
+					klog.InfoS("Updated PacketCapture", "PacketCapture", klog.KObj(pc), "capturedPacketsNum", captureState.capturedPacketsNum)
+				}
+			}
+		case <-timer.C:
+			pc, err := c.packetCaptureLister.Get(captureState.name)
+			if err != nil {
+				return fmt.Errorf("get PacketCapture failed: %w", err)
+			}
+			klog.InfoS("PacketCapture timeout", "name", pc.Name)
+			err = errors.New(captureTimeoutReason)
+			if updateErr := c.updatePacketCaptureStatus(pc.Name, captureState.capturedPacketsNum, "", err); updateErr != nil {
+				klog.ErrorS(updateErr, "The PacketCapture timed out, but PacketCapture status update failed", "name", pc.Name)
+			}
+			return err
+		}
+	}
+}
+
+func (c *Controller) getPodIP(podRef *crdv1alpha1.PodReference) (net.IP, error) {
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(podRef.Name, podRef.Namespace)
+	var result net.IP
+	if len(podInterfaces) > 0 {
+		result = podInterfaces[0].GetIPv4Addr()
+	} else {
+		pod, err := c.kubeClient.CoreV1().Pods(podRef.Namespace).Get(context.TODO(), podRef.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to get Pod %s/%s: %w", podRef.Namespace, podRef.Name, err)
+		}
+		podIPs := make([]net.IP, len(pod.Status.PodIPs))
+		for i, ip := range pod.Status.PodIPs {
+			podIPs[i] = net.ParseIP(ip.IP)
+		}
+		result = util.GetIPv4Addr(podIPs)
+	}
+	if result == nil {
+		return nil, fmt.Errorf("cannot find IP with IPV4 AddressFamily for Pod %s/%s", podRef.Namespace, podRef.Name)
+	}
+	return result, nil
+}
+
+func (c *Controller) createMatchPacket(pc *crdv1alpha1.PacketCapture) (*binding.Packet, error) {
```

Comment:
The process that converts `crdv1alpha1.PacketCapture` to `binding.Packet`, then to `[]bpf.Instruction` doesn't seem necessary. Why does it the intermediate data which is specific to Openflow?

Reply:
The `openflow.Packet` provides a good struct definition to store the parsed data. eg. From `source.pod` or `source.ip` -> `sourceIP`, `tcp.srcPort` or `udp.srcPort` -> `srcPort`.  Will replace this with a custom struct.

Reply:
removed in the new interface.